src/baldr/curler.cc

```diff
@@ -0,0 +1,123 @@
+#include "baldr/curler.h"
+#include "midgard/logging.h"
+
+#include <stdexcept>
+
+#ifdef CURL_STATICLIB
+
+#if defined(_MSC_VER) && !defined(NOGDI)
+#define NOGDI // prevents winsock2.h drag in wingdi.h
+#endif
+
+#include <curl/curl.h>
+
+#if defined(_MSC_VER) && defined(GetNameInfo)
+#undef GetNameInfo // winsock2.h imports #define GetNameInfo which clashes with
+                   // EdgeInfo::GetNameInfo
+#endif
+
+namespace {
+
+struct curl_singleton_t {
+  curl_singleton_t() {
+    curl_global_init(CURL_GLOBAL_DEFAULT);
+  }
+  ~curl_singleton_t() {
+    curl_global_cleanup();
+  }
+};
+
+static std::shared_ptr<CURL> init_curl() {
+  static curl_singleton_t s;
+  return std::shared_ptr<CURL>(curl_easy_init(), [](CURL* c) { curl_easy_cleanup(c); });
+}
+
+char ALL_ENCODINGS[] = "";
+
+size_t write_callback(char* in, size_t block_size, size_t blocks, std::vector<char>* out) {
+  if (!out) {
+    return static_cast<size_t>(0);
+  }
+  out->insert(out->end(), in, in + (block_size * blocks));
+  return block_size * blocks;
+}
+
+} // namespace
+
+namespace valhalla {
+namespace baldr {
+
+struct curler_t::pimpl_t {
```

Comment:
i moved the use of curl into a pimpl because it was the only way to hide curl into a single compilation unit in an effective way. curl isnt friendly to forward declarations which would have made this much easier.